File: tests/test_advanced.py
import random
from algoritmo.busqueda_binaria import busqueda_binaria
from framework.hybrid_framework import TestRunner

def registrar_pruebas_avanzadas(runner: TestRunner):
    
    # 1. Contratos
    def test_contratos():
        def it_validar_tipos():
            runner.expect(lambda: busqueda_binaria("no_es_lista", 10)).to_raise(TypeError)
            runner.expect(lambda: busqueda_binaria([1, 2, 3], "no_es_entero")).to_raise(TypeError)
        runner.it("debe validar precondicion de tipos de parametros (TypeError)", it_validar_tipos)
        
        def it_validar_orden():
            runner.expect(lambda: busqueda_binaria([5, 3, 8, 1], 3)).to_raise(ValueError)
        runner.it("debe validar precondicion de ordenamiento de lista (ValueError)", it_validar_orden)
        
        def it_validar_postcondiciones():
            runner.expect(busqueda_binaria([10, 20, 30], 20)).to_be(1)
            runner.expect(busqueda_binaria([10, 20, 30], 25)).to_be(-1)
        runner.it("debe validar postcondiciones de exito en busquedas", it_validar_postcondiciones)
        
    runner.describe("Búsqueda Binaria - Pruebas de Contratos", test_contratos)

    # 2. Propiedades
    def test_propiedades():
        def it_verificar_invariantes():
            for _ in range(50):
                tamano = random.randint(1, 100)
                datos = sorted(random.sample(range(-1000, 1000), tamano))
                
                # Caso A: Buscar elemento que existe
                objetivo_existente = random.choice(datos)
                indice = busqueda_binaria(datos, objetivo_existente)
                runner.expect(datos[indice]).to_be(objetivo_existente)
                
                # Caso B: Buscar elemento que posiblemente no existe
                objetivo_inexistente = random.randint(-2000, 2000)
                indice_b = busqueda_binaria(datos, objetivo_inexistente)
                if indice_b != -1:
                    runner.expect(datos[indice_b]).to_be(objetivo_inexistente)
                else:
                    runner.expect(objetivo_inexistente not in datos).to_be(True)
        runner.it("debe verificar invariantes con datos autogenerados", it_verificar_invariantes)
        
    runner.describe("Búsqueda Binaria - Property-based Testing", test_propiedades)

    # 3. Mocks y Espías
    def test_mocks():
        def it_registrar_espias():
            spy = runner.create_spy(return_value=123)
            resultado = spy("parametro_de_prueba", 456)
            runner.expect(resultado).to_be(123)
            runner.expect(spy.get_call_count()).to_be(1)
            runner.expect(spy.was_called_with("parametro_de_prueba", 456)).to_be(True)
        runner.it("debe registrar correctamente llamadas e interactuar con spies", it_registrar_espias)
        
    runner.describe("Búsqueda Binaria - Mocks y Espías", test_mocks)

File: tests/testAdvanced.js
const { busquedaBinaria } = require('../algoritmo/busquedaBinaria.js');

function registrarPruebasAvanzadas(runner) {
    // 1. Bloque de Contratos
    runner.describe("Búsqueda Binaria - Pruebas de Contratos", () => {
        runner.it("debe validar precondicion de tipos de parametros (TypeError)", () => {
            runner.expect(() => busquedaBinaria("no_es_arreglo", 10)).toThrow(TypeError);
            runner.expect(() => busquedaBinaria([1, 2, 3], "no_es_entero")).toThrow(TypeError);
        });

        runner.it("debe validar precondicion de ordenamiento de lista (Error)", () => {
            runner.expect(() => busquedaBinaria([5, 3, 8, 1], 3)).toThrow(Error);
        });

        runner.it("debe validar postcondiciones de exito en busquedas", () => {
            runner.expect(busquedaBinaria([10, 20, 30], 20)).toBe(1);
            runner.expect(busquedaBinaria([10, 20, 30], 25)).toBe(-1);
        });
    });

    // 2. Bloque de Propiedades (Property-based)
    runner.describe("Búsqueda Binaria - Property-based Testing", () => {
        runner.it("debe verificar invariantes con datos autogenerados", () => {
            for (let i = 0; i < 50; i++) {
                const tamano = Math.floor(Math.random() * 100) + 1;
                // Generar números ordenados aleatorios únicos
                const conjunto = new Set();
                while (conjunto.size < tamano) {
                    conjunto.add(Math.floor(Math.random() * 2000) - 1000);
                }
                const datos = Array.from(conjunto).sort((a, b) => a - b);

                // Caso A: Buscar elemento que sí existe
                const objetivoExistente = datos[Math.floor(Math.random() * datos.length)];
                const indice = busquedaBinaria(datos, objetivoExistente);
                runner.expect(datos[indice]).toBe(objetivoExistente);

                // Caso B: Buscar elemento aleatorio (puede no existir)
                const objetivoInexistente = Math.floor(Math.random() * 4000) - 2000;
                const indiceB = busquedaBinaria(datos, objetivoInexistente);
                if (indiceB !== -1) {
                    runner.expect(datos[indiceB]).toBe(objetivoInexistente);
                } else {
                    runner.expect(datos.includes(objetivoInexistente)).toBe(false);
                }
            }
        });
    });

    // 3. Mocks y Espías
    runner.describe("Búsqueda Binaria - Mocks y Espías", () => {
        runner.it("debe registrar correctamente llamadas e interactuar con spies", () => {
            const spy = runner.createSpy(123);
            const resultado = spy("parametro_de_prueba", 456);

            runner.expect(resultado).toBe(123);
            runner.expect(spy.getCallCount()).toBe(1);
            runner.expect(spy.wasCalledWith("parametro_de_prueba", 456)).toBe(true);
        });
    });
}

module.exports = { registrarPruebasAvanzadas };
